src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala

```diff
@@ -52,71 +79,166 @@ case class BlockHeader(
     */
   lazy val hash: ByteString = ByteString(kec256(this.toBytes: Array[Byte]))
 
-  lazy val hashAsHexString: String = Hex.toHexString(hash.toArray)
-
-  val hasCheckpoint: Boolean = checkpoint.isDefined
+  lazy val hashAsHexString: String = ByteStringUtils.hash2string(hash)
 
   def idTag: String =
     s"$number: $hashAsHexString"
 }
 
 object BlockHeader {
 
-  import Checkpoint._
-  import io.iohk.ethereum.rlp.RLPImplicitConversions._
-  import io.iohk.ethereum.rlp.RLPImplicits._
-
-  private implicit val checkpointOptionDecoder = implicitly[RLPDecoder[Option[Checkpoint]]]
-  private implicit val checkpointOptionEncoder = implicitly[RLPEncoder[Option[Checkpoint]]]
-
   val emptyOmmerHash = ByteString(Hex.decode("1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347"))
 
   def getEncodedWithoutNonce(blockHeader: BlockHeader): Array[Byte] = {
-    val rlpEncoded = blockHeader.toRLPEncodable match {
-      case rlpList: RLPList if blockHeader.checkpoint.isDefined =>
-        // post ECIP1098 & ECIP1097 block
-        val rlpItemsWithoutNonce = rlpList.items.dropRight(4) ++ rlpList.items.takeRight(2)
-        RLPList(rlpItemsWithoutNonce: _*)
-
-      case rlpList: RLPList if blockHeader.treasuryOptOut.isDefined =>
-        // Post ECIP1098 block without checkpoint
-        val rlpItemsWithoutNonce = rlpList.items.dropRight(3) :+ rlpList.items.last
-        RLPList(rlpItemsWithoutNonce: _*)
-
-      case rlpList: RLPList if blockHeader.treasuryOptOut.isEmpty =>
-        // Pre ECIP1098 & ECIP1097 block
-        RLPList(rlpList.items.dropRight(2): _*)
-
-      case _ => throw new Exception("BlockHeader cannot be encoded without nonce and mixHash")
+    // toRLPEncodeable is guaranteed to return a RLPList
+    val rlpList: RLPList = blockHeader.toRLPEncodable.asInstanceOf[RLPList]
+
+    val rlpItemsWithoutNonce = blockHeader.extraFields match {
+      case HefPostEcip1097(_ , _) =>
+        // Post ECIP1098 & ECIP1097 block
+        rlpList.items.dropRight(4) ++ rlpList.items.takeRight(2)
+
+      case HefPostEcip1098(_) =>
+        // Post ECIP1098 block, Pre ECIP1097
+        rlpList.items.dropRight(3) :+ rlpList.items.last
+
+      case HefPreEcip1098 =>
+        // Pre ECIP1098 and ECIP1097 block, encoding works as if optOut and checkpoint fields weren't defined for backwards compatibility
+        rlpList.items.dropRight(2)
     }
-    rlpEncode(rlpEncoded)
+    rlpEncode(RLPList(rlpItemsWithoutNonce: _*))
   }
 
+  // scalastyle:off parameter.number
+  def buildPreECIP1098Header(parentHash: ByteString,
```

Comment:
Is it worth adding all these methods? Maybe we could just have a default param `extraFields: HeaderExtraFields = HefEmpty` on the default constructor? And as for the extra fields, I don't think wrapping them in the HEF-constructor is that tedious to necessitate these methods

Reply:
I found them useful when changing the block header structure as that required less changes. Given that and that we might eventually further change it (in case we want 3 different header structures) I saw no reason to delete them once I had them

But I don't have too much of a strong opinions right now, if you have stronger opinions on deleting them I can do so

Reply:
Well, I think that using a default arg for `extraFields` would result in even fewer changes. Any future changes should also require less maintenance. Do you think there is another advantage to having these factories?

Reply:
Tbh I don't like much default arg in general (exceptuating some use cases) as I see them too error prone... but given the subjectivity of this and that the PR has been open for a long time I'll change this